core/tee/tee_sql_fs.c

```diff
@@ -0,0 +1,1321 @@
+	struct sql_fs_fd *fdp = NULL;
+	int rflags = flags | TEE_FS_O_RDWR; /* Need to read/write meta */
+	int exists;
+	int fd = -1;
+
+	DMSG("(file: %s, flags: %d)...", file, flags);
+
+	exists = !sql_fs_access_rpc(file, TEE_FS_F_OK);
+	if (flags & TEE_FS_O_CREATE) {
+		if ((flags & TEE_FS_O_EXCL) && exists) {
+			*errno = TEE_ERROR_ACCESS_CONFLICT;
+			goto exit;
+		}
+	} else {
+		if (!exists) {
+			*errno = TEE_ERROR_ITEM_NOT_FOUND;
+			goto exit;
+		}
+	}
+
+	fdp = (struct sql_fs_fd *)calloc(1, sizeof(*fdp));
+	if (!fdp) {
+		*errno = TEE_ERROR_OUT_OF_MEMORY;
+		goto exit;
+	}
+
+	fdp->fd = sql_fs_open_rpc(errno, file, rflags);
+	if (fdp->fd < 0)
+		goto exit;
+
+	fdp->flags = flags;
+
+	fd = read_meta(errno, fdp);
+	if (fd < 0)
+		goto exit;
+
+	fd = handle_get(&fs_db, fdp);
+
+exit:
+	if (fd < 0)
+		free(fdp);
+	DMSG("...%d", fd);
+	return fd;
+}
+
+static int sql_fs_read(TEE_Result *errno, int fd, void *buf, size_t len)
+{
+	struct sql_fs_fd *fdp = handle_lookup(&fs_db, fd);
+	size_t remain_bytes = len;
+	uint8_t *data_ptr = buf;
+	uint8_t *block = NULL;
+	int start_block_num;
+	int end_block_num;
+	int res = -1;
+	int ret;
+
+	DMSG("(fd: %d, buf: %p, len: %zu)...", fd, (void *)buf, len);
+
+	if (!fdp) {
+		*errno = TEE_ERROR_BAD_PARAMETERS;
+		goto exit;
+	}
+
+	if (fdp->pos + len > fdp->meta.length)
+		len = fdp->meta.length - fdp->pos;
+
+	if (!len) {
+		res = 0;
+		goto exit;
+	}
+
+	if (!buf) {
+		*errno = TEE_ERROR_BAD_PARAMETERS;
+		goto exit;
+	}
+
+	if (fdp->flags & TEE_FS_O_WRONLY) {
+		*errno = TEE_ERROR_ACCESS_CONFLICT;
+		goto exit;
+	}
+
+	start_block_num = block_num(fdp->pos);
+	end_block_num = block_num(fdp->pos + len + 1);
+
+	block = malloc(BLOCK_SIZE);
+	if (!block) {
+		*errno = TEE_ERROR_OUT_OF_MEMORY;
+		goto exit;
+	}
+
+	sql_fs_begin_transaction_rpc();
+
+	while (start_block_num <= end_block_num) {
+		tee_fs_off_t offset = fdp->pos % BLOCK_SIZE;
+		size_t size_to_read = MIN(remain_bytes, (size_t)BLOCK_SIZE);
+
+		if (size_to_read + offset > BLOCK_SIZE)
+			size_to_read = BLOCK_SIZE - offset;
+
+		/*
+		 * REVISIT: implement read_block_partial() since we have
+		 * write_block_partial()?
+		 */
+		res = read_block(errno, fdp, start_block_num, block);
+		if (res < 0)
+			goto exit;
+
+		memcpy(data_ptr, block + offset, size_to_read);
+
+		data_ptr += size_to_read;
+		remain_bytes -= size_to_read;
+		fdp->pos += size_to_read;
+
+		start_block_num++;
+	}
+	res = 0;
+exit:
+	sql_fs_end_transaction_rpc(res < 0);
+	free(block);
+	ret = (res < 0) ? res : (int)len;
+	DMSG("...%d", ret);
+	return ret;
+}
+
+static int sql_fs_write(TEE_Result *errno, int fd, const void *buf, size_t len)
+{
+	struct sql_fs_fd *fdp = handle_lookup(&fs_db, fd);
+	size_t remain_bytes = len;
+	const uint8_t *data_ptr = buf;
+	tee_fs_off_t orig_pos;
+	int start_block_num;
+	int end_block_num;
+	int res = -1;
+	int ret;
+
+	DMSG("(fd: %d, buf: %p, len: %zu)...", fd, (void *)buf, len);
+
+	if (!fdp) {
+		*errno = TEE_ERROR_BAD_PARAMETERS;
+		goto exit;
```

Comment:
I think the call will do nothing if not currently in a transaction. But you're right, this is not good style. I will change this to jump after `sql_fs_end_transaction_rpc()` instead. And same in other places.